src/v/redpanda/admin/server.cc

```diff
@@ -746,6 +746,12 @@ ss::future<ss::httpd::redirect_exception> admin_server::redirect_to_leader(
     // - if the host in the Host header matches one of our advertised kafka
     //   addresses, then assume that the peer's advertised kafka address
     //   with the same index will also be their public admin API address.
+    //   - a match in this case might be an exact match or the appearance
+    //     that the advertised address is a subdomain of request hostname.
+    //     this allows redirection of requests made through a headless service
+    //     in a k8s environment, but is not intended to cover all such cases,
+    //     since the service hostname could be anything. Context:
+    //     https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-hostname-and-subdomain-fields
```

Comment:
The reason why I think it is in scope is because it is an alternative way to implement this PR. We could avoid subdomain matching altogether if we fall back to the `advertised_rpc_api` instead of the `rpc_server`. `advertised_rpc_api` (or more specifically `advertised_rpc_api.value_or(rpc_server)`) has the nice property that it is used by the [brokers to connect to each other](https://github.com/redpanda-data/redpanda/blob/ceae89a87e3a901bece8ef848409dd9f922f4652/src/v/cluster/cluster_utils.cc#L52-L54), so we know it is routable internally. It's a simpler fix because it makes no assumption on the request host structure, and it extends trivially to the case when someone sends a request to the cluster by IP. So I think it's the preferable alternative to this PR. What do you think?

Reply:
Is it a good guess that an arbitrary client with a route to the headless server will _also_ have a route to the advertised rpc address? Could that, for example, circumvent a load balancer or something?

Reply:
Yeah, that's a fair point. Afaik, with our supported kubernetes charts, both the advertised rpc address and the headless service address are routable within the same domain (within the kubernetes cluster), and they should take the same path since the headless service domain effectively resolves to the same IP as the advertised rpc address of one of the nodes. But there's nothing stopping anyone from overwriting the advertised rpc address config...